src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,573 @@
+        (sampleInfo, numMismatches)
+      }
+      .toList.sortBy(_._2).take(2) match {
+        case Nil                              => (this.unmatchedSample, Int.MaxValue, Int.MaxValue)
+        case List(bestTuple)                  => (bestTuple._1, bestTuple._2, Int.MaxValue)
+        case List(bestTuple, secondBestTuple) => (bestTuple._1, bestTuple._2, secondBestTuple._2)
+      }
+    }
+    else {
+      (this.unmatchedSample, Int.MaxValue, Int.MaxValue)
+    }
+
+    // Make sure we are within the parameter limits and update barcode metrics if necessary
+    if (maxMismatches < bestMismatches || maxNoCalls < numNoCalls || (secondBestMismatches - bestMismatches) < minMismatchDelta) {
+      (this.unmatchedSample, Int.MaxValue)
+    }
+    else {
+      (bestSampleInfo, bestMismatches)
+    }
+  }
+
+  /** Demultiplexes a given set of reads from the same template.  The same number of reads should be given as read
+    * structures.
+    *
+    * The sample barcoded bases from each read are extracted and concatenated in the same order as the given reads. They
+    * are matched against the sample barcode bases for each sample.
+    * */
+  def demultiplex(reads: FastqRecord*): DemuxRecord = {
+    require(reads.nonEmpty, "No reads given for demultiplexing.")
+    require(reads.length == expectedNumberOfReads, s"Expected '$expectedNumberOfReads' number of reads but found '${reads.length}'.")
+
+    // Get the sample
+    val allBases                    = reads.map(_.bases)
+    val (sampleInfo, numMismatches) = matchSampleBarcode(allBases)
+
+    // Create the SAMRecords.
+    val header            = sampleInfo.header
+    val sampleId          = sampleInfo.rg.getId
+    val molecularBarcode  = this.molecularBarcode(allBases, "-")
+    val sampleBarcode     = this.sampleBarcode(allBases, "-")
+    var readNumber        = 1
+    val records           = reads.zip(readStructures).flatMap { case (read, readStructure) =>
+      val templateSegments = readStructure.template
+      if (templateSegments.isEmpty) {
+        None
+      }
+      else {
+        // NB: strict=false in case folks have trimmed the reads prior can be handled gracefully.
+        val subReads: Seq[SubRead] = ReadStructure.structureReadWithQualities(bases=read.bases, qualities=read.quals, segments=templateSegments, strict=false)
+        val bases                  = subReads.map(_.bases).mkString
+        val quals                  = subReads.flatMap(_.quals).mkString
+        require(bases.length == quals.length, s"Bases and qualities have differing lengths for read: ${read.header}")
+
+        // Build the [[SAMRecord]]
+        val record  = new SAMRecord(header)
+        record.setReadName(read.name)
+        record.setReadString(bases)
+        record.setReadUnmappedFlag(true)
+        if (pairedEnd) {
+          record.setReadPairedFlag(true)
+          record.setMateUnmappedFlag(true)
+          if (readNumber == 1) {
```

Comment:
This could be:
```scala
readNumber match {
  case 1 => record.setFirstOfPairFlag(true)
  case 2 => record.setSecondOfPairFlag(true) 
  case n => throw new IllegalStateException(s"Invalid read number '$n' for read: ${read.header}")
}
```